File: leetcode_alg/solution61.cpp
//
// Created by wangheng on 2022/8/27.
//
#include <iostream>
#include "listnode.h"

using namespace std;

class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (head == nullptr) {
            return head;
        }
        ListNode *p = head;
        ListNode *end = nullptr;
        int n = 0;
        while (p != nullptr) {
            end = p;
            p = p->next;
            n++;
        }
        k = k % n;
        if (k == 0) {
            return head;
        }
        ListNode *fast = head, *slow = head;
        while (k > 0) {
            fast = fast->next;
            k--;
        }
        while (fast != end) {
            fast = fast->next;
            slow = slow->next;
        }
        ListNode *tmp = slow->next;
        slow->next = nullptr;
        fast->next = head;
        return tmp;
    }
};

void testSolution() {
    auto s = new Solution;
    ListNode *root = new ListNode(1, new ListNode(2, new ListNode(3,
                                                                  new ListNode(4, new ListNode(5)))));
    auto pNode = s->rotateRight(root, 3);
    print_list(pNode);
    free_list(root);
    delete s;
}

int main() {
    testSolution();
    return 0;
}

File: golang/solution/solution61.go
package solution

import "fmt"

// 旋转链表
func rotateRight(head *ListNode, k int) *ListNode {

	var (
		p          *ListNode = head
		listLength int       = 0
		realK      int       = k
		fast       *ListNode = head
		slow       *ListNode = head
	)
	for p != nil {
		listLength++
		p = p.Next
	}
	if listLength == 0 {
		return head
	}
	realK = k % listLength

	if realK == 0 {
		return head
	}

	// 快慢指针，快指针先走realK步，然后慢指针和快指针一起往后走，直到快指针到达链表末尾
	for i := 0; i < realK; i++ {
		fast = fast.Next
	}

	for fast.Next != nil {
		slow = slow.Next
		fast = fast.Next
	}
	fast.Next = head
	head = slow.Next
	slow.Next = nil
	return head
}

func Solution61() {
	// head := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{5, nil}}}}}
	var head *ListNode = nil
	head = rotateRight(head, 0)
	for head != nil {
		fmt.Println(head.Val)
		head = head.Next
	}
}
